src/channel/chat.js

```diff
@@ -335,6 +336,9 @@ ChatModule.prototype.processChatMsg = function (user, data) {
         }
     }
 
+    if(!coolholePostProcessChatMessage(this.channel, user, data, msgobj))
+        return;
```

Comment:
Speaking of mutating arguments, I generally think that sucks haha. I know sync does that everywhere but IMO just returning a new object to merge in to the `msgobj` seems must easier to handle.

Reply:
> question: do we really care what `coolholePostProcessChatMessage` returns? And do we want to end this method early if something goes tits up?
> 
> If our intention is to have regular cytube functionality "just werk" if our code fails, what our methods return probably doesn't matter as long as we don't mutate arguments

I coded it this way because of other coolhole features that eventually need to be created. It is to provide a way of "denying" the chat message from going through, its mainly for stuff like the cp points allow/deny the chat message if they have enough points (like /secretary, /danmu, /highlight). I could have just called "coolholePostProcessChatMessage()" without the if check and return, but i was just looking ahead i guess.


> Speaking of mutating arguments, I generally think that sucks haha. I know sync does that everywhere but IMO just returning a new object to merge in to the `msgobj` seems must easier to handle.

Well, i was returning "true/false" for the allow/deny functionality. I was also trying to minimize the # of code changes made to cytube files. If we just do something like "msgobj = coolholePostProcessChatMessage(...)", its essentially accomplishing the same thing as mutating the arguments itself. Unless we want to do something like:
`
results = coolholePostProcessChatMessage(this.channel, user, data, msgobj);
if(results.allow)
    msgobj = results.msgobj
else
    return;
`
I was just trying minimize the logic/code on the cytube files thats all.

Reply:
Okay, my mistake. I looked at the code again for current coolhole. Disregard my last comment
In current coolhole prod, all the features that deny the chat message (/secretary, /danmu, /highlight) are actually handled in the coolpoints module by simply returning early and not calling "processChatMessage()" again.

So, my code is redundant, my bad. I think I was trying to cram too much in my brain while doing it and I thought "oh yeah, we'll definitely need this for later" and just got messed up. I'll remove the if statement.

As for the "mutating argument", the handleDebtEffects() also needs to be called modify the text, and we basically modified the msgobj directly in coolhole prod. Though we could do a similar thing there, just do a `msgobj.msg = coolholeHandleDebtEffects(...);" or something like that.

Maybe you're right, and maybe I'm trying to do too much. I'll change it so only my "coolholeMeta" object is attached to msgobj in chat and see how that goes.

Reply:
>It is to provide a way of "denying" the chat message from going through, its mainly for stuff like the cp points allow/deny the chat message if they have enough points (like /secretary, /danmu, /highlight)

Oh I see. We don't want to emit a chatmsg event if they have insufficient funds. Then yea I think you're right. We'd probably wanna do something like you suggested
```js
try {
  const coolholeMeta = coolholePostProcessChatMessage(this.channel, user, data, msgobj);
  if (coolholeMeta.allow)
    msgObj.meta.coolholeMeta = coolholeMeta;
  else
    // either return or do our socket emit logic here... probably better to stuff this server-mod
    user.socket.emit("coolholeDisallow", {
      action: coolholeMeta.error.action,
      message: coolholeMeta.error.message
    }
 }
 catch (Execption e) {
   // Probably wanna log more than just exception 
   console.error("Some shit happened in coolhole", e);
   this.channel.logger.log("Some shit happened in coolhole" + JSON.stringify(e));
   // Ignoring uncaught exception
}
```


>In current coolhole prod, all the features that deny the chat message (/secretary, /danmu, /highlight) are actually handled in the coolpoints module by simply returning early and not calling "processChatMessage()" again.

Yeah I think that _sucks_ haha. It was a naive approach that just assumed nothing would ever fail. I'm trying to advocate for something like a try catch around our code in `chat.js` and just make sure we leave the `msgObj` unmolested if it blows up. This way, regular chat functionality should just work if our code blows up. IE: I'd rather see `/danmu hey` in the chat rather than nothing appearing with no feedback.

>I'll change it so only my "coolholeMeta" object is attached to msgobj in chat and see how that goes.

Yeah this is really what I was trying to say. This way we don't, say, mutate the object partially, have coolhole code fail, and then the message object be partially updated.
It's also generally pretty painful to think about how the object was mutated in a function call instead of just thinking what a function returns.

Reply:
>It is to provide a way of "denying" the chat message from going through, its mainly for stuff like the cp points allow/deny the chat message if they have enough points (like /secretary, /danmu, /highlight)

Oh I see. We don't want to emit a chatmsg event if they have insufficient funds. Then yea I think you're right. We'd probably wanna do something like you suggested
```js
try {
  const coolholeMeta = coolholePostProcessChatMessage(this.channel, user, data, msgobj);
  if (coolholeMeta.allow)
    msgObj.meta.coolholeMeta = coolholeMeta;
  else
    // break the chat message flow; CH should emit its own events and only its own events
    return;
    /* probably better to stuff this server-mod
    user.socket.emit("coolholeDisallow", {
      action: coolholeMeta.error.action,
      message: coolholeMeta.error.message
    }
    */
 }
 catch (Execption e) {
   // Probably wanna log more than just exception 
   console.error("Some shit happened in coolhole", e);
   this.channel.logger.log("Some shit happened in coolhole" + JSON.stringify(e));
   // Ignoring uncaught exception
}
```

>In current coolhole prod, all the features that deny the chat message (/secretary, /danmu, /highlight) are actually handled in the coolpoints module by simply returning early and not calling "processChatMessage()" again.

Yeah that was "smart" of me I guess since we have to override what the server would emit when sending a message. If we wanted to avoid modifying the chatMsg client side code entirely, we'd need a way to fork from the usual sockets called just below this. I think what you were suggesting is probably the best approach.

>I'll change it so only my "coolholeMeta" object is attached to msgobj in chat and see how that goes.

Yeah this is really what I was trying to say. This way we don't, say, mutate the object partially, have coolhole code fail, and then the message object be partially updated. IE: I'd rather see `/danmu hey` in the chat rather than nothing appearing with no feedback. 

It's also generally pretty painful to think about how the object was mutated in a function call instead of just thinking what a function returns.